clippy2021.md

```diff
@@ -0,0 +1,178 @@
+implement with a guarantee of no/few false positives (FPs). This won't change.
+But we should consider implementing more lint groups or improve the process of
+classifying lints. This will also involve internal changes, and will be
+addressed later in this document again.
+
+[#5537](https://github.com/rust-lang/rust-clippy/issues/5537)
+[#6366](https://github.com/rust-lang/rust-clippy/issues/6366)
+
+#### `lints.toml` configuration
+
+This is something that comes up every now or then. A configuration file, where
+lint levels can be defined and reused across projects. Discussions about this
+often lead to nothing specific/actionable or to "we need an RFC for this". And
+this is exactly what we need to do. Get together with the cargo team and write
+an RFC and implement such a configuration file somehow and somewhere.
+
+[#3164](https://github.com/rust-lang/rust-clippy/issues/3164)
+[cargo#5034](https://github.com/rust-lang/cargo/issues/5034)
+[IRLO](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135/8)
+
+
+#### FP rate
+
+In the worst case, new lints are only available in nightly for 2 weeks, before
+hitting beta and ultimately stable. This and the fact that fewer people use
+nightly Rust nowadays makes it more probable that a lint with many FPs hits
+stable. This leads to annoyed users, that will disable these new lints in the
+best case and to more annoyed users, that will stop using Clippy in the worst.
+We should think of a process to prevent this from happening. This will also be
+addressed later in this document.
+
+[#6429](https://github.com/rust-lang/rust-clippy/issues/6429)
+
+#### Clippy book
+
+We should consider writing a Clippy book, like other projects have. This will
+then be the central place where everything you need to know to use Clippy is
+collected in a structured way.
+
+[#6011](https://github.com/rust-lang/rust-clippy/issues/6011)
+
+### Reliability
+
+We should make sure to make Clippy more reliable and stable for our users
+
+#### FP rate
+
+In the worst case, new lints are only available in nightly for 2 weeks, before
+hitting beta and ultimately stable. This and the fact that fewer people use
+nightly Rust nowadays makes it more more probably that a lint with many FPs hits
+stable. This leads to annoyed users, that will disable these new lints in the
+best case and to more annoyed users, that will stop using Clippy in the worst.
+We should think of a process to prevent this from happening. This will also be
+addressed later in this issue.
+
+[#6429](https://github.com/rust-lang/rust-clippy/issues/6429)
+
+## Internal
+
+(The end of) 2020 has shown, that Clippy has to think about the available
+resources, especially regarding management and maintenance of the project. This
+section address issues affecting team members and contributors.
+
+### Management
+
+In 2020 we got over 1000 open issues with now between 25-35 open PRs. This is
```

Comment:
From my perspective this is simultaneously a win and a loss 😄 . Clippy's maintainer set is particularly robust and can handle people dropping out pretty easily, so much that 3-4 people temporarily scaling back involvement did not actually cause the project to grind to a halt! But still, ideally we're as robust as possible!

Since most of our issues are lint suggestions I do not mind there being 1k lints 😄 , that's just 1k ideas for folks to implement. The PR backlog is worrying however.

Either way, I think it's _always_ a good idea to grow maintainership and processes. I do think there's a lot better we could do here, and excited to see what we can come up with!

Reply:
> Clippy's maintainer set is particularly robust and can handle people dropping out pretty easily

That is correct. This year was just a bit unlucky because besides the people who dropped out, the remaining maintainers (@Manishearth, @ebroto and I) had to deal with IRL stuff, which collided with hacktoberfest, which was unlucky. We're still dealing with the backlog, I think. More people would make things more stable, so I like to establish a process for growing the team.

> Since most of our issues are lint suggestions I do not mind there being 1k lints

Are they? I don't know. I have the feeling that bug reports are on the rise. This may also be something we want to look at. Which lint has how many issues? How many issues are new lint suggestions? Generally getting some metrics to track the state Clippy is in.

Reply:
>This may also be something we want to look at. Which lint has how many issues? How many issues are new lint suggestions?

I agree. I also think we lack a process for deciding when a lint should be downgraded. How many false positives we accept for a warn/deny-by-default lint? And for a pedantic lint?

A good amount of our issues are quite old. IMO we should go through the whole issue list and re-triage, closing issues that don't apply anymore, and getting a better idea of lints that maybe should be downgraded.